### PR TITLE
[SPARK-21746][SQL]there is an java.lang.IllegalArgumentException when the filter contains nondeterminate expressions

### PR DESCRIPTION
**Pull request by heary-cao**
## What changes were proposed in this pull request?
Currently, We do interpretedpredicate optimization, but not very well, because when our filter contained an nondeterminate expression.
in spark-shell. execute the following SQL statement:
```
val path = "/home/spark/files/7"
Seq(1 -> "a").toDF("a", "b").write.partitionBy("a").parquet(path)
val df = spark.read.parquet(path)
df.filter(rand(10) <= 1.0).select($"a").show
```
Spark throws an exceptions  java.lang.IllegalArgumentException:
```
java.lang.IllegalArgumentException: requirement failed: Nondeterministic expression org.apache.spark.sql.catalyst.expressions.Rand should be initialized before eval.
  at scala.Predef$.require(Predef.scala:224)
  at org.apache.spark.sql.catalyst.expressions.Nondeterministic$class.eval(Expression.scala:291)
  at org.apache.spark.sql.catalyst.expressions.RDG.eval(randomExpressions.scala:34)
  at org.apache.spark.sql.catalyst.expressions.BinaryExpression.eval(Expression.scala:415)
  at org.apache.spark.sql.catalyst.expressions.InterpretedPredicate.eval(predicates.scala:38)
  at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex$$anonfun$10.apply(PartitioningAwareFileIndex.scala:180)
  at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex$$anonfun$10.apply(PartitioningAwareFileIndex.scala:179)
  at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
  at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
  at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
  at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex.prunePartitions(PartitioningAwareFileIndex.scala:179)
  at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex.listFiles(PartitioningAwareFileIndex.scala:64)
  at org.apache.spark.sql.execution.FileSourceScanExec.org$apache$spark$sql$execution$FileSourceScanExec$$selectedPartitions$lzycompute(DataSourceScanExec.scala:180)
  at org.apache.spark.sql.execution.FileSourceScanExec.org$apache$spark$sql$execution$FileSourceScanExec$$selectedPartitions(DataSourceScanExec.scala:177)
  at org.apache.spark.sql.execution.FileSourceScanExec$$anonfun$21.apply(DataSourceScanExec.scala:279)
  at org.apache.spark.sql.execution.FileSourceScanExec$$anonfun$21.apply(DataSourceScanExec.scala:278)
  at scala.Option.map(Option.scala:146)
  at org.apache.spark.sql.execution.FileSourceScanExec.<init>(DataSourceScanExec.scala:278)
  at org.apache.spark.sql.execution.datasources.FileSourceStrategy$.apply(FileSourceStrategy.scala:106)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:63)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:63)
  at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
  at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
  at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:93)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2$$anonfun$apply$2.apply(QueryPlanner.scala:78)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2$$anonfun$apply$2.apply(QueryPlanner.scala:75)
  at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
  at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
  at scala.collection.Iterator$class.foreach(Iterator.scala:893)
  at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
  at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
  at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1336)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2.apply(QueryPlanner.scala:75)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2.apply(QueryPlanner.scala:67)
  at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
  at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:93)
  at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:84)
  at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:80)
  at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:89)
  at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:89)
  at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3031)
  at org.apache.spark.sql.Dataset.head(Dataset.scala:2344)
  at org.apache.spark.sql.Dataset.take(Dataset.scala:2557)
  at org.apache.spark.sql.Dataset.showString(Dataset.scala:241)
  at org.apache.spark.sql.Dataset.show(Dataset.scala:671)
  at org.apache.spark.sql.Dataset.show(Dataset.scala:630)
  at org.apache.spark.sql.Dataset.show(Dataset.scala:639)
```
This PR describes solving this problem by adding the initialize method in InterpretedPredicate.

## How was this patch tested?

Should be covered existing test cases and add new test cases.

### CODE REVIEW

**Comment by heary-cao** [comment]
cc @gatorsmile  @cloud-fan

**Comment by gatorsmile** [comment]
ok to test

**Comment by heary-cao** [comment]
@cloud-fan @gatorsmile 
Could you take a look again?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -2029,4 +2029,15 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
       testData2.select(lit(7), 'a, 'b).orderBy(lit(1), lit(2), lit(3)),
       Seq(Row(7, 1, 1), Row(7, 1, 2), Row(7, 2, 1), Row(7, 2, 2), Row(7, 3, 1), Row(7, 3, 2)))
   }
+
+  test("SPARK-21746: nondeterministic expressions correctly for filter predicates") {
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Do we need to turn off this? It looks irrelevant to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 2]
```diff
@@ -31,11 +31,22 @@ object InterpretedPredicate {
   def create(expression: Expression, inputSchema: Seq[Attribute]): InterpretedPredicate =
     create(BindReferences.bindReference(expression, inputSchema))
 
-  def create(expression: Expression): InterpretedPredicate = new InterpretedPredicate(expression)
+  def create(expression: Expression): InterpretedPredicate = {
+    val predicate = new InterpretedPredicate(expression)
+    predicate.initialize(0)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: `0` is not a `partitionIndex` and  `0` will go to `n.initialize(0)` in line 46. Is it correct?

**Comment by gatorsmile** [comment]
Will review it tonight.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 1 above)
Comment (heary-cao) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: aha,you are right.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 2 above)
Comment (heary-cao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: we not need InterpretedPredicate.initialize and modify _object InterpretedPredicate.create_
```
  def create(expression: Expression): InterpretedPredicate = {
    expression.foreach {
      case n: Nondeterministic => n.initialize(0)
      case _ =>
    }
    new InterpretedPredicate(expression)
  }
```

Comment (heary-cao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: The consideration then was the consistency of class InterpretedPredicate, so add initialize method of the class InterpretedPredicate.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 3]
```diff
@@ -2029,4 +2029,13 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
       testData2.select(lit(7), 'a, 'b).orderBy(lit(1), lit(2), lit(3)),
       Seq(Row(7, 1, 1), Row(7, 1, 2), Row(7, 2, 1), Row(7, 2, 2), Row(7, 3, 1), Row(7, 3, 2)))
   }
+
+  test("SPARK-21746: nondeterministic expressions correctly for filter predicates") {
+    withTempPath { path =>
+      val p = path.getAbsolutePath
+      Seq(1 -> "a").toDF("a", "b").write.partitionBy("a").parquet(p)
+      val df = spark.read.parquet(p)
+      checkAnswer(df.filter(rand(10) <= 1.0).select($"a"), Row(1))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: why this test case will trigger `InterpretedPredicate`?

Comment (heary-cao) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: I think that predicates is not empty.

Comment (heary-cao) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: In addition, I tried to validate the spark 2.0.2 version, and it won't trigger  InterpretedPredicate Exception. 
spark 2.0.2  _InterpretedPredicate.create_
```
  def create(expression: Expression): (InternalRow => Boolean) = {
    expression.foreach {
      case n: Nondeterministic => n.setInitialValues()
      case _ =>
    }
    (r: InternalRow) => expression.eval(r).asInstanceOf[Boolean]
  }
```

**Comment by heary-cao** [comment]
@dongjoon-hyun @cloud-fan Do you have any suggestions?

**Comment by cloud-fan** [comment]
Good catch! I think we do need `InterpretedPredicate.initialize`, but I don't think we should call `initialize` in `InterpretedPredicate.create`. Can you check out where we create `InterpretedPredicate` and call `initialize` there?

**Comment by heary-cao** [comment]
@cloud-fan, Thank you for your suggest. I found 4 calls to InterpretedPredicate.create and modify it. Can you take a look again if you have time?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 4]
```diff
@@ -362,7 +362,9 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
       str
     }
     logWarning(s"Codegen disabled for this expression:\n $logMessage")
-    InterpretedPredicate.create(expression, inputSchema)
+    val predicate = InterpretedPredicate.create(expression, inputSchema)
+    predicate.initialize(0)
+    predicate
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: The partition index is not always 0.

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/SimpleTextRelation.scala [hunk 5]
```diff
@@ -90,6 +90,7 @@ class SimpleTextSource extends TextBasedFileFormat with DataSourceRegister {
         }.reduceOption(expressions.And).getOrElse(Literal(true))
         InterpretedPredicate.create(filterCondition, inputAttributes)
       }
+      predicate.initialize(0)
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/SimpleTextRelation.scala]: The predicate here can only be `sources.GreaterThan`. It can't include non-deterministic expressions.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: I ran this test without any exception. Are you sure this test can reproduce this issue?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 4 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: I checked the few places calling `newPredicate`. They are either already do `initialize` properly, or don't need to do it.

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/SimpleTextRelation.scala (hunk 5 above)
Comment (heary-cao) [reply sql/hive/src/test/scala/org/apache/spark/sql/sources/SimpleTextRelation.scala]: thanks， modify it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 4 above)
Comment (heary-cao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: thanks, remove it.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: First, you don't add any comment to explain it. Reviewers don't know that.
Second, a new test should test against some regressions or the proposed changes. This test is neither of both.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: The test you add should trigger `InterpretedPredicate` as @cloud-fan mentioned. It should hit the code path you modified in this PR, not depending on another PR.

Comment (heary-cao) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: sorry，I update the description of PR. thanks.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: can this test expose the bug?

Comment (heary-cao) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: I'm updating to the latest version for validation.

Comment (heary-cao) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: @viirya, @cloud-fan 
This trigger condition is associated with #18918. It will be more prone to this exception
The current spark master branch does not trigger this code path.
put this change on #18918 and close this PR.
thanks.

**Comment by heary-cao** [comment]
Leave a comment

**Comment by gatorsmile** [comment]
Any reason you closed it?

**Comment by gatorsmile** [comment]
You do not need to open multiple PRs for the same issue.

**Comment by heary-cao** [comment]
@gatorsmile , This should be a problem for code execution, semantics, and consistency.
This trigger condition: _FileSourceScanExec.partitionFilters_  is not null and contains nondeterministic function. however, the current spark master branch does not trigger this code path.  because _PhysicalOperation_ excluded that the condition is nondeterministic of Filter.
Similar spark sql: ` SELECT t1.i3 from tableorc t1 where rand(10) <= 0.5`
 the current spark master branch  executed Plan:
```
*Project [i3#0]
+- *Filter (rand(10) <= 0.5)
   +- *FileScan orc default.tableorc[i3#0,i4#1,i5#2,i6#3,i7#4,i8#5,i9#6,i10#7,i11#8,s2#9,s3#10,s4#11,s5#12,s6#13,s7#14,d2#15,d3#16,d4#17,d5#18,d6#19,d7#20,i2#21] Batched: false, Format: ORC, Location: CatalogFileIndex[file:/home/cxw/spark/bin/spark-warehouse/tableorc], PartitionCount: 2, PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i3:int,i4:int,i5:int,i6:int,i7:int,i8:int,i9:int,i10:int,i11:int,s2:string,s3:string,s4:st...
```
 PartitionFilters is **[]**

Thus, this trigger condition is associated with #18918. it will be more prone to this exception.
#18918 PR executed Plan:
```
*Project [i3#0]
+- *Filter (rand(10) <= 0.5)
   +- *FileScan orc default.tableorc[i3#0,i2#21] Batched: false, Format: ORC, Location: PrunedInMemoryFileIndex[file:/home/cxw/spark/bin/spark-warehouse/tableorc/i2=0], PartitionCount: 1, PartitionFilters: [(rand(10) <= 0.5)], PushedFilters: [], ReadSchema: struct<i3:int>
```
PartitionFilters is **[(rand(10) <= 0.5)]**
thanks.
